share/lib/python/neuron/tests/utils/__init__.py

```diff
@@ -0,0 +1,87 @@
+"""
+Utilities for writing tests
+"""
+from contextlib import contextmanager
+
+
+@contextmanager
+def cvode_enabled(enabled):
+    """
+    Helper for tests to enable/disable CVode, leaving the setting as they found it.
+
+    For example:
+    from neuron.tests.utils import cvode_enabled
+    with cvode_enabled(True):
+        # cvode is enabled here
+        some_test_using_cvode()
+    # cvode is disabled again
+    """
+    from neuron import h
+
+    old_status = h.cvode_active()
+    h.cvode_active(enabled)
+    try:
+        yield None
+    finally:
+        h.cvode_active(old_status)
+
+
+@contextmanager
+def cvode_use_long_double(enabled):
+    """
+    Helper for tests to enable/disable long double support in CVode, leaving the setting as they found it.
+    """
+    from neuron import h
+
+    cv = h.CVode()
+    old_setting = cv.use_long_double()
+    cv.use_long_double(enabled)
+    if enabled != old_setting:
+        # FIXME find a better way of forcing an already-initialised CVode to take the new use_long_double setting into account
+        cv.atol(cv.atol())
+    try:
+        yield None
```

Comment:
Yes; `contextmanager` is a decorator from `contextlib` (https://docs.python.org/3/library/contextlib.html#contextlib.contextmanager) that makes it easy to implement context managers for Python's `with` statements.

In this case, it allows us to write something like:
```python
# use_long_double has some value X
with cvode_use_long_double(True):
  # use_long_double is True
  do_work()
# use_long_double has value X again
```
which I think is a good pattern to adopt in the tests we launch from `pytest`, because it helps to automate cleanup. If we do things manually, we are prone to having different tests affect each other via NEURON's global state. (e.g. `test_b.py` runs with whatever number of NrnThreads that `test_a.py` left over).

The `None` value is just because we're not taking a value it, as is done in other contexts when some resource is being managed, for example:
```python
with open("foo.txt") as my_file:
  do_stuff(my_file)
# my_file is closed
```
if you did
```python
with cvode_use_long_double(True) as foo:
  # foo is None, because of the yield None statement
```
you would get the `None` value, but I don't see any especially meaningful value to put there in this case.